Player tracking api

i509VCB opened the pull request:
Supersedes #517. Splits off the player tracking aspect of v0 networking into a new module. Also adds events that are called when an entity starts or stops being tracked by a player.

`PlayerStream` isn't great tbf. A better name would be nice, possibly just going for `PlayerTracking`?

JamiesWhiteShirt commented:
We should also consider being consistent with "watching" vs. "tracking", as this API uses both. This is more of a problem in Yarn than anything, but this API can be consistent.

My opinions:
- I find "tracking" to be the most appropriate term.
- If we should preserve the term "player stream", it should be called `PlayerStreams`, not `PlayerStream`. But yes, it's not great because the term is too generic. +1 for `PlayerTracking`.

i509VCB commented:
I've decided on "tracking" for the naming purposes of this api.

JamiesWhiteShirt commented on fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/package-info.java:
```diff
@@ -14,12 +14,12 @@
+ * The fabric networking player tracking API, version 1.
+ *
+ * <p>This API allows mods to easily select player recipients of packets
+ * based on distance from a position, their watching status of server
```
```suggestion
 * based on distance from a position, their tracking status of server
```
Missed one.

NinjaPhenix commented on fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/PlayerTracking.java:
```diff
@@ -0,0 +1,180 @@
+	 * @param world the server world
+	 * @param pos   the block position
+	 * @return the players tracking the block position
+	 * @throws IllegalArgumentException if the world is not a server world
```
Is this true? I only see NullPointerExceptions, also this requires an instance of ServerWorld ( or subclass )

i509VCB replied:
Those javadoc throws tags are not removed since they don't apply any more.

Player3324 commented on fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/PlayerTracking.java:
```diff
@@ -0,0 +1,182 @@
+ *
+ * <p>These methods should only be called on the server thread and only be used on logical a server.
+ */
+public final class PlayerTracking {
```
The name isn't great, particularly when combined with some methods, like "PlayerTracking.all" or "PlayerTracking.tracking". Maybe something like PlayerLookup?

Player3324 commented on fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/PlayerTracking.java:
```diff
@@ -0,0 +1,182 @@
+	 * @param pos   the chunk in question
+	 * @return the players tracking the chunk
+	 */
+	public static Stream<ServerPlayerEntity> tracking(ServerWorld world, ChunkPos pos) {
```
There is some inconsistency between Stream and Collection use. I'd use Collection everywhere since partial consumption is unlikely and it's the simpler API.

Player3324 reviewed (commented):
Splitting such a small and stable API off is IMO a bad idea, each module comes with some complexity and overhead cost. The events are fine, the player lookup helper could use a few minor tweaks.

i509VCB committed:
Split out player finding stuff to networking player tracking API v1

Signed-off-by: liach <liach@users.noreply.github.com>

Forward v0 PlayerStream to new module, add entity track events


Rename module to player tracking


Well javadoc can make sense


Decide on tracking for the name


Update fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/package-info.java

Co-authored-by: Erlend Åmdal <erlend@aamdal.com>
Remove exceptions from javadoc that are not thrown


javadoc fix again


Handle a case where the player manager happens to be null


rename player tracking to player lookup


Yeet

i509VCB commented:
Closing since this is being cherry-picked into #1081
